### PR TITLE
`render_docs(autolink = TRUE)` uses `downlit`

### PR DESCRIPTION
**Pull request by vincentarelbundock**
This PR adds an `autolink` argument to `render_docs()` to process vignettes, man pages, and home page using `downlit`. This will auto-link function calls to web-based documentation.

Links to external functions work without further intervention. Links to an `altdoc` website requires users to add a `pkgdown.yml` file in the root of the website like this:

```yaml
urls:
  reference: https://vincentarelbundock.github.io/tinytable/man
  article: https://vincentarelbundock.github.io/tinytable/vignettes
```

Users should probably host this file in `altdoc`

Warning: `downlit` will only recognize internal links if `pkgdown.yml` *is already hosted live on the website*. This means that this won't work the first time the website is generated, and that local updates will not be reflected in links.

TODO:

+ [x] Create `altdoc/pkgdown.yml` automatically with `setup_docs()`
+ [x] Check docute
+ [x] Check docsifiy
+ [x] Check mkdocs
+ [x] **How do we do unit tests for this? The tricky part is it requires a `pkgdown.yml` hosted on a live website.**

### CODE REVIEW

**Comment by vincentarelbundock** [comment]
@etiennebacher I think this is ready for review now.

In the end, I decided to set `autolink=FALSE` by default for reasons explained in this part of the new documentation:

> Auto-link:
> 
>      When the ‘autolink’ argument is ‘TRUE’, ‘altdoc’ will use the
>      ‘downlit’ package to replace the function names on the package
>      website by links to web-based package documentation. This works
>      for base R libraries and any package published on CRAN.
> 
>      To allow internal links to functions documented by ‘altdoc’, we
>      need to include links to correct URLs in the ‘altdoc/pkgdown.yml’
>      file. By default, this file is populated with links to the first
>      URL in the ‘DESCRIPTION’. When calling
>      ‘render_docs(autolink=TRUE)’, the ‘pkgdown.yml’ file is moved to
>      the root of the website.
> 
>      Importantly, ‘downlit’ requires the ‘pkgdown.yml’ to be live on
>      the website to create links. This means that links will generally
>      not be updated when making purely local changes. Also, links may
>      not be updated the first time an ‘altdoc’ website is published to
>      the web.
> 
>      Note that the ‘autolink’ argument works best for Quarto-based
>      websites. ‘mkdocs’ appears to ignore ‘downlit’ annotations
>      altogether. ‘docute’ and ‘docsify’ display ‘downlit’ annotations,
>      but CSS styling and code highlighting sometimes suffer.
> 
>

**Comment by vincentarelbundock** [comment]
In case you want to see it "live", I updated one of my websites: https://marginaleffects.com/vignettes/predictions.html

**Review by etiennebacher — COMMENTED** [review]
Oh yeah that looks great, nice one! Too bad it doesn't work with `mkdocs` but the example with `marginaleffects` looks really good.

I have some comments about some code refactor but I'm fine with the functionality. I think it's good to keep it opt-in for now if it messes up the formatting for `docsify` and `docute`

File: R/autolink.R [hunk 1]
```diff
@@ -0,0 +1,30 @@
+.autolink <- function(path = ".") {
+    # everyting wrapped in tryCatch() because especially error prone with weird pandoc memory errors
+    html_files <- c(tryCatch(fs::dir_ls(fs::path_join(c(path, "docs/vignettes")), regexp = "\\.html$"), error = function(e) NULL),
+                    tryCatch(fs::dir_ls(fs::path_join(c(path, "docs/man")), regexp = "\\.html$"), error = function(e) NULL))
+    for (h in html_files) {
+        tmp <- try(downlit::downlit_html_path(h, h), silent = TRUE)
```
[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply R/autolink.R:6]: I don't think that's what it means. `downlit_html_node()` just returns an object for interactive use in R. In my experiences, I think I saw no changes when `downlit_html_path()` failed.

File: R/autolink.R [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            cli::cli_alert_danger(sprintf("Failed to auto-link: %s", h))
+        }
+
+    }
+    # h <- fs::path_join(c(path, "docs/index.html"))
+    # if (fs::file_exists(h)) {
+    #     downlit::downlit_html_path(h, h)
+    # }
+
+    md_files <- c(
```
[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply R/autolink.R:13]: We need this to work in docsify and docute, where there are no HTML files.

File: R/autolink.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (etiennebacher) [reply R/autolink.R:6]: Oh right, I think I misread it

File: inst/gha/altdoc.yaml [hunk 3]
```diff
@@ -61,7 +61,7 @@ jobs:
           install.packages(".", repos = NULL, type = "source")
           install.packages("pkgload")
           pkgload::load_all()
-          altdoc::render_docs(parallel = FALSE, freeze = FALSE)
+          altdoc::render_docs(parallel = TRUE, freeze = FALSE)
```
[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply inst/gha/altdoc.yaml:64]: Probably a mistake. I think this should be FALSE by default unless it's a really big website and the user has thought serious about resources available on github actions.

File: R/utils.R [hunk 4]
```diff
@@ -127,6 +127,26 @@
...
 
 
+.add_pkgdown <- function(path = ".") {
+  if (!isTRUE(.dir_is_package(path))) {
+    stop(".add_pkgdown() must be run from the root of a package.", call. = FALSE)
+  }
+  url <- desc::desc_get_urls()
+  fn <- fs::path_join(c(path, "altdoc/pkgdown.yml"))
+  if (!fs::file_exists(fn) && length(url) > 0) {
+    url <- url[1]
```
[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply R/utils.R:137]: 4a75c2b4a

File: R/autolink.R [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    for (m in md_files) {
+        tmp <- try(downlit::downlit_md_path(m, m), silent = TRUE)
+        if (inherits(tmp, "try-error")) {
+            cli::cli_alert_danger(sprintf("Failed to auto-link: %s", h))
+        }
+    }
+    # m <- fs::path_join(c(path, "docs/README.md"))
+    # if (fs::file_exists(m)) {
+    #     downlit::downlit_html_path(m, m)
+    # }
```
[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply R/autolink.R]: 4a75c2b4a

**Review by etiennebacher — APPROVED** [review]
Thanks, feel free to merge if you're done

**Pull request merged**